### PR TITLE
[#4] [Integrate] As a user, I can sign in with email and password

### LINKED ISSUE #4: [Integrate] As a user, I can sign in with email and password
## Why

Before a user can use the app, they have to sign in with their email and password, so we need a Sign In screen to let the user authenticate to the app

## Acceptance Criteria

- Email or Password TextField is emptied -> disable the Login button
- When tapping on Login Button, show the loading indicator and call the login API
  - If successful, hide the loading indicator and navigate to the Home screen
  - If failure, show a native alert with the error message from the response

## Design

Sign in

Link: https://www.figma.com/file/GjRPOjDyZ6f4EDL3wKarRK/Challenge---Mobile-App?node-id=31%3A195

### PR DESCRIPTION
**Pull request by markgravity**
https://github.com/markgravity/kmm-ic/issues/4

## What happened 👀

- Setup Navigator
- Integrate the Login screen
- Write Unit Test for `LoginViewModel`
- Update Sourcery templates to generate mocks for UseCase

## Insight 📝

After several tries, we cannot combine the configuration of `AutoMockable` & `AutoSharedUsecase` into one `.sourcery.yml` only, so we end up creating a new `.shared-sourcery.yml` to contain the configuration of `AutoSharedUsecase` then after triggering a build in Xcode
- Run a build script to generate the `AutoSharedUseCase.generated.swift`
- Run the normal Sourcery build script to generate the `AutoMockable.generated.swift`

## Proof Of Work 📹

https://user-images.githubusercontent.com/17875522/233429732-16e931cd-eba5-473c-9384-d77bebb307cd.mp4

### CODE REVIEW

File: ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    init() {
+        Publishers.CombineLatest($email, $password)
+            .map { !$0.0.isEmpty && !$0.1.isEmpty }
+            .assign(to: &$isAllValidated)
+    }
+
+    func login() {
+        isLoading = true
+
+        loginUseCase.callAsPublisher(email: email, password: password)
```
Comment (suho) [inline ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift]: Why don't use `callAsFunction`

File: ios/tools/sourcery/templates/Shared.stencil [hunk 2]
```diff
@@ -4,6 +4,7 @@ import Combine
 
 {% for type in types.protocols where type.name | hasSuffix:"UseCase" %}
   {% for method in type.methods where method.name | hasPrefix:"callAsNative(" %}
+// sourcery:begin: AutoMockable
```
Comment (suho) [inline ios/tools/sourcery/templates/Shared.stencil:7]: Please fix all changes in one PR #67. Or maybe separate a chore for setting up a new template/stencil for this one

**Review by suho — CHANGES_REQUESTED** [review]

File: ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift (hunk 1 above)
Comment (markgravity) [reply ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift]: @suho `callAsFunction` is quite confusing for me, what is the call as function mean? 🤔 
`callAsNative` is also call it as a function

File: ios/tools/sourcery/templates/Shared.stencil (hunk 2 above)
Comment (markgravity) [reply ios/tools/sourcery/templates/Shared.stencil:7]: @suho updated at [b5c86d3](https://github.com/markgravity/kmm-ic/pull/67/commits/b5c86d34cf43b9e1cac1bdf1dc2ae0f36ddf159a)

File: ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift (hunk 1 above)
Comment (suho) [reply ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift]: [callAsFunction](https://github.com/apple/swift-evolution/blob/main/proposals/0253-callable.md) is a feature of Swift. With that, you can use use case like this:
```swift
loginUseCase(email: email, password: password)
```

File: ios/.swiftlint.yml [hunk 3]
```diff
@@ -7,6 +7,7 @@ excluded:
   - KMMICTests/Sources/Mocks/Sourcery/AutoMockable.generated.swift
   - KMMIC/Sources/Supports/Helpers/Rswift/R.generated.swift
   - KMMIC/Sources/Supports/Helpers/Shared/Shared.generated.swift
+  - KMMIC/Sources/Supports/Helpers/Shared/Shared.tmp.generated.swift
```
Comment (suho) [inline ios/.swiftlint.yml]: How can we generate this? 🤔

File: ios/KMMICTests/Sources/Dummy/AuthToken+Dummy.swift [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+//  Copyright © 2023 Nimble. All rights reserved.
+//
+
+import Shared
+
+@testable import KMMIC
+
+extension AuthToken {
+
+    static let dummy = AuthToken(
```
Comment (suho) [inline ios/KMMICTests/Sources/Dummy/AuthToken+Dummy.swift:15]: You might need to use `AutoDummyable`: https://github.com/nimblehq/swift-auto-dummy-ios-initiative

File: ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift (hunk 1 above)
Comment (markgravity) [reply ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift]: @suho please ignore this naming for now, I will open a PR to renaming all of them at once 🙏

File: ios/KMMICTests/Sources/Dummy/AuthToken+Dummy.swift (hunk 4 above)
Comment (markgravity) [reply ios/KMMICTests/Sources/Dummy/AuthToken+Dummy.swift:15]: @suho I tried, but it doesn't work with the model from KMM 🙏

File: ios/.swiftlint.yml (hunk 3 above)
Comment (markgravity) [reply ios/.swiftlint.yml]: We generate it in a build script
CleanShot 2023-04-24 at 15 10 09

File: ios/KMMIC/Sources/Supports/Extensions/SwiftUI/View+Alert.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+//
+//  Created by MarkG on 20/04/2023.
+//  Copyright © 2023 Nimble. All rights reserved.
+//
+
+import Foundation
+import Shared
+import SwiftUI
+
+struct AlertDescription: Identifiable, Equatable {
```
Comment (phongvhd93) [inline ios/KMMIC/Sources/Supports/Extensions/SwiftUI/View+Alert.swift]: We could separate this into another file

File: ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Injected(\.loginUseCase) private var loginUseCase
+
+    private var cancellableBag = Set<AnyCancellable>()
+
+    @Published var email: String = ""
+    @Published var password: String = ""
+    @Published var isAllValidated: Bool = false
+    @Published var isLoading: Bool = false
+    @Published var didLogin: Bool = false
```
Comment (phongvhd93) [inline ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift]: We don't have to set variable type for these properties

Comment (markgravity) [reply ios/KMMIC/Sources/Presentation/Modules/LoginScreen/LoginViewModel.swift]: @phongvhd93 updated at [d74bf92](https://github.com/markgravity/kmm-ic/pull/69/commits/d74bf9208aaaaeda1de406a04b09cde4dfd9f9f1)

File: ios/KMMICTests/Sources/Dummy/AuthToken+Dummy.swift (hunk 4 above)
Comment (markgravity) [reply ios/KMMICTests/Sources/Dummy/AuthToken+Dummy.swift:15]: @suho FYI, after debugging the Sourcery is unable to get the `init` function from a model in Kotlin

File: ios/KMMIC/Sources/Supports/Extensions/SwiftUI/View+Alert.swift (hunk 5 above)
Comment (markgravity) [reply ios/KMMIC/Sources/Supports/Extensions/SwiftUI/View+Alert.swift]: @phongvhd93 updated at [d74bf92](https://github.com/markgravity/kmm-ic/pull/69/commits/d74bf9208aaaaeda1de406a04b09cde4dfd9f9f1)

**Review by phongvhd93 — APPROVED** [review]

File: ios/.swiftlint.yml (hunk 3 above)
Comment (markgravity) [reply ios/.swiftlint.yml]: @suho More information in the `Insight` section
```
After several tries, we cannot combine the configuration of `AutoMockable` & `AutoSharedUsecase` into one `.sourcery.yml` only, so we end up creating a new `.shared-sourcery.yml` to contain the configuration of `AutoSharedUsecase` then after triggering a build in Xcode

- Run a build script to generate the AutoSharedUseCase.generated.swift
- Run the normal Sourcery build script to generate the AutoMockable.generated.swift
```

File: ios/KMMIC/Sources/Application/App.swift [hunk 7]
```diff
@@ -7,17 +7,20 @@
...
 import Shared
 import SwiftUI
+import Combine
 
 @main
 struct KMMApp: App {
 
     @UIApplicationDelegateAdaptor private var appDelegate: AppDelegate
-
+    var subscription: AnyCancellable?
```
Comment (suho) [inline ios/KMMIC/Sources/Application/App.swift]: What is this property?

Comment (markgravity) [reply ios/KMMIC/Sources/Application/App.swift]: @suho update at [02e7eba](https://github.com/markgravity/kmm-ic/pull/69/commits/02e7eba009d5903ccdc7fb1710b65b03bafacbab) to remove the test code 🙈

**Review by suho — APPROVED** [review]

**Pull request merged**